core/src/main/kotlin/io/customer/android/core/di/DiGraph.kt

```diff
@@ -1,48 +1,111 @@
 package io.customer.android.core.di
 
 abstract class DiGraph {
+    /**
+     * Map of dependencies that can be overridden with mocks in test functions.
+     */
     val overrides: MutableMap<String, Any> = mutableMapOf()
 
     /**
-     * Call when you want to override a dependency with a mock in test functions.
-     * `di.overrideDependency(CustomerIOInstance::class.java, customerIOMock)`
-     *
-     * See `CustomerIOTest` (or any other class that is top-level of the module) for an example on use.
+     * Map of singleton instances of dependencies in the graph.
+     */
+    val singletons: MutableMap<String, Any> = mutableMapOf()
+
+    /**
+     * Internal function to get an overridden dependency in the graph.
+     * This function is marked as public only because it is inlined in other public functions.
+     * Do not call this function directly. Instead, use the `newInstance` or `singleton` functions
+     * to get the dependency.
      */
-    fun <DEP> overrideDependency(dependency: Class<DEP>, value: DEP) {
-        overrides[dependency.simpleName] = value as Any
+    inline fun <reified Dependency : Any> getOverriddenInstance(): Dependency? {
+        return overrides[Dependency::class.java.name] as? Dependency
     }
 
     /**
-     * Call in Digraph property getters to get allow mock overriding before trying to get actual instance.
-     * ```
-     * val fileStorage: FileStorage
-     *   get() = override() ?: FileStorage(siteId, context)
-     * ```
-     * Note: Don't forget to include the interface in the property declaration:
+     * Internal function to get or create a singleton instance of a dependency in the graph.
+     * This function is thread-safe and will only create a single instance of the dependency.
+     * This function is marked as public only because it is inlined in other public functions.
+     * Do not call this function directly. Instead, use the `singleton` function.
+     */
+    inline fun <reified Dependency : Any> getOrCreateSingletonInstance(
+        newInstanceCreator: () -> Dependency
+    ): Dependency {
+        // Use a synchronized block to prevent multiple threads from creating multiple instances of the singleton.
+        synchronized(lock = this) {
```

Comment:
we can get rid of this with suggestion above

Reply:
Responded above